### PR TITLE
JS: Loosen the source in `js/shell-command-constructed-from-input`

### PR DESCRIPTION
**Pull request by erik-krogh**
Gets a TP/TN for CVE-2021-23326

Three changes: 
 - Flag library input in sub-packages to better support mono-repos where the "sub-packages" are published independently. 
 - Look for a main module in `index.js` and `src/index.js`, instead of just the former.  
 - Add a link to the source in the alert message (I used it while debugging backwards flow, and I decided to keep it).

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/NodeModuleResolutionImpl.qll [hunk 1]
```diff
@@ -96,8 +96,11 @@ File resolveMainModule(PackageJSON pkg, int priority) {
     )
   )
   or
-  result =
-    tryExtensions(pkg.getFile().getParentContainer(), "index", priority - prioritiesPerCandidate())
+  exists(Folder folder | folder = pkg.getFile().getParentContainer() |
+    result =
+      tryExtensions([folder, folder.getChildContainer("src")], "index",
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/NodeModuleResolutionImpl.qll:101]: It is quite common to name the `src` folder `lib` in JavaScript projects. But perhaps that is not the case with `lib/index.js`?

File: javascript/ql/test/query-tests/Security/CWE-078/lib/subLib/index.js [hunk 2]
```diff
@@ -1,7 +1,7 @@
 var cp = require("child_process")
 
 module.exports = function (name) {
-	cp.exec("rm -rf " + name); // OK - this file belongs in a sub-"module", and is not the primary exported module.
```
Comment (esbena) [inline javascript/ql/test/query-tests/Security/CWE-078/lib/subLib/index.js:4]: Hmm. We explicitly decided not to flag this kind of pattern earlier. Should we do a run-on-all to ensure that we are not forgetting something?

File: javascript/ql/src/semmle/javascript/PackageExports.qll [hunk 3]
```diff
@@ -17,33 +17,12 @@ DataFlow::ParameterNode getALibraryInputParameter() {
...
-}
-
-/**
- * Gets a value exported by the main module from one of the topmost `package.json` files (see `getTopmostPackageJSON`).
+ * Gets a value exported by the main module from a named `package.json` file.
  * The value is either directly the `module.exports` value, a nested property of `module.exports`, or a method on an exported class.
  */
 private DataFlow::Node getAValueExportedByPackage() {
-  result = getAnExportFromModule(getTopmostPackageJSON().getMainModule())
+  result =
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/PackageExports.qll:24]: 🎉 this is a nice simplification! Let's hope we don't need it.

**Review by esbena — COMMENTED** [review]
I have some concerns about the FPs we may introduce. But you are the authority on those.

File: javascript/ql/src/semmle/javascript/NodeModuleResolutionImpl.qll (hunk 1 above)
Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/NodeModuleResolutionImpl.qll:101]: [`lib`](https://github.com/search?q=extension%3Ajs+path%3Alib%2Findex&type=Code&ref=advsearch&l=&l=) is less common than [`src`](https://github.com/search?q=extension%3Ajs+path%3Asrc%2Findex&type=Code&ref=advsearch&l=&l=).   
It seems reasonable to add it, so I'll do that.

File: javascript/ql/test/query-tests/Security/CWE-078/lib/subLib/index.js (hunk 2 above)
Comment (erik-krogh) [reply javascript/ql/test/query-tests/Security/CWE-078/lib/subLib/index.js:4]: > Hmm. We explicitly decided not to flag this kind of pattern earlier. Should we do a run-on-all to ensure that we are not forgetting something?

Lets do that.  

I think the reason I added the filter was from seeing noise in various sub-packages.  
Often these sub-packages were just copy-pasted third party libraries, or only used internally.   

But they're not FPs, they're just benign results (and people should stop copy-pasting code).  

I'm running [an evaluation](https://github.com/dsp-testing/erik-krogh-dca/tree/run/topPack-default-libaryInputs/reports) now, lets wait for that before we do a run-on-all.

**Comment by erik-krogh** [comment]
[Evaluation came back](https://github.com/dsp-testing/erik-krogh-dca/tree/run/topPack-default-libaryInputs/reports), and there's something to look at (new results). 

I think the performance regression is spurious. 
I only evaluated with two queries, and I couldn't replicate the biggest regression locally (the relevant predicates didn't even show up in the clause timing reports).

**Comment by erik-krogh** [comment]
> [Evaluation came back](https://github.com/dsp-testing/erik-krogh-dca/tree/run/topPack-default-libaryInputs/reports), and there's something to look at (new results).

The `unsafe-shell-command` results are TPs from copy-pasted code-bases (the package.json points to another repo).  

And the `poly-redos` results are mostly from various sub-packages.  
Most of the ones I inspected looked like they could be published independently.  

So I think it looks safe.  
But a run-on-all would likely be a good idea. 
(Waiting for https://github.com/github/codeql/pull/5449 before I do a run-on-all).

**Comment by erik-krogh** [comment]
> But a run-on-all would likely be a good idea.
> (Waiting for #5449 before I do a run-on-all).

Run on all is done: https://github.com/github/codeql/pull/5449#issuecomment-806080470   
The FP rate looks OK.

**Review by esbena — APPROVED** [review]

**Pull request merged**